### ISSUE TITLE
RFI: Please add comments to ipset entries

### ISSUE THREAD

**Issue by TriMoon** [t-woerner/firewalld#194]
It would be **much** appreciated if the firewalld UI, in particular (and possibly ``firewall-cmd``),  would allow us to add comments to entries in ipsets.

**Comment by leonbloy** [comment]
Related https://serverfault.com/questions/893112/migrating-from-iptables-to-firewalld-commenting-rules

**Comment by rozmansi** [comment]
I too am planning a migration from _/etc/sysconfig/iptables_ to FirewallD and would love the support for commenting: entries in ipsets, particular rich rules, etc.

Could `firewall-cmd` be upgraded to just copy the `<!-- -->` elements when updating the XML config files?

**Comment by erig0** [comment]
> Could firewall-cmd be upgraded to just copy the  elements when updating the XML config files?

Possibly. I think it would be useful for all the XML files.

In the mean time you could use the ipset description to stash metadata. e.g.

    # firewall-cmd --permanent --ipset=foobar --set-description="10.1.1.0/24: lab -- 10.1.2.0/24: office"

```
# cat /etc/firewalld/ipsets/foobar.xml 
<?xml version="1.0" encoding="utf-8"?>
<ipset type="hash:ip">
  <description>10.1.1.0/24: lab -- 10.1.2.0/24: office</description>
  <entry>10.1.1.0/24</entry>
  <entry>10.1.2.0/24</entry>
</ipset>
```

**Comment by FrenchFranck** [comment]
Accord the change log history, since V6.20 the comment are supported for each entry 
ipset create foo hash:ip comment
ipset add foo 192.168.1.1/24 comment "This is my comment 2"
ipset add foo 192.168.1.2/24 comment "This is my comment 2"

But some Linux distribution does not support this option (see https://bugzilla.redhat.com/show_bug.cgi?id=1496859)
On my Fedora 28 its ok, but with an error message appear whitout importance.

Use description is not sufficient. As SpareSimian suggested, it will be a good idea to define the comment as an attribute of each entry. Like this :
.......
<entry comment="This is my comment 1">192.168.1.1</entry>
<entry comment="This is my comment 2">192.168.10.5</entry>
......


But how to specify the option to indicate the ipset have to use the comment 
 ipset create foo hash:ip **comment**

**Comment by judchuks** [comment]
www. aidenpromotions. com/articles

That is one way in which you could drop your articles online for free without charge. Go there and post your articles without paying a dime!

**Comment by TriMoon** [comment]
> In the mean time you could use the ipset description to store some metadata. e.g.
The OP was not only about commenting the ipset as a group, but more like each entry inside the ipset.
Eg. each element put inside the group...

**Comment by lacrimal** [comment]
Hi,
There is not much info since last message in this topic.
So is it possible to add `comment ` option to ipsets, rich-rules etc. ?

**Comment by lasseoe** [comment]
Any update on this? It'd be super useful to be able to add comments to at least rich rules. Some may argue that comments (documentation) should be part of a configuration management system, but while debugging on a host it's A LOT easier with a basic comment on a rule.
Cheers :-)

**Comment by moontide** [comment]
Same requirement here.

If entry list getting large, it's hard to track the purpose of each entry.

**Comment by Ruppsn** [comment]
Same requirement here.

No Solution found yet.

**Comment by swimlappy** [comment]
+1 for this. Would be very helpful.

**Comment by Ruppsn** [comment]
We now use a workaround for comments:

Example rich rule
```
  <rule family="ipv4">
    <source address="XXX.XX.XXX.X/24"/>
    <port port="YYYY" protocol="tcp"/>
    <log prefix="Log prefix used as comment " level="info">
      <limit value="2/d"/>
    </log>
    <accept/>
  </rule>
```

with the limit value there will be 2 log entries per day.

not perfect but better than having no idea why a rule is there

**Comment by iTrooz** [comment]
Would the project accept a PR to add this ?

I'd implement this using the existing `--set-short` and `--set-description` flags for rules. e.g.:
`firewall-cmd --permanent --add-port 22/tcp --set-description="Internal SSH"`

the classic (long) description would be shown when using `--query-port` (with or without `--get-description`) (e.g. `firewall-cmd --query-port 50001/tcp`)

The short description would be shown when showing rule overview. For example, `firewall-cmd --list-all` would print `  ports: 50001/tcp (internal website), 50002/tcp (something else)`

What do you think about this ?
Ping @erig0 for confirmation before I start working on it

**Comment by tonyjonespdx** [comment]
In regard to previous comment.     Issue 489 which specifically mentioned comments for --add-port was marked a duplicate of this but now you're saying you won't address --add-port in regard to this issue.

**Comment by ArthurRM** [comment]
Isn't regular XML comments already supported? For my installation this seems to work fine:

```
<?xml version="1.0" encoding="utf-8"?>
<ipset type="hash:ip">
  ...
  <!-- vulnerability scanning; too many requests -->
  <entry>80.94.95.29</entry>
  <!-- user-agent and query parameter injection -->
  <entry>145.79.184.115</entry>
</ipset>

```

**Comment by erig0** [comment]
> Isn't regular XML comments already supported? For my installation this seems to work fine:

Yes. However, if you update the ipset via firewall-cmd, then the comments will be lost. Firewalld doesn't parse and rewrite the comments. It could though. Just needs someone to add support to the XML readers/writers.

**Comment by L34T** [comment]
> to add support to the XML readers/writers.

imo, adding XML comment support would create a new UX problem: once you have around 20 manually added bans with comments - removing a specific ban would require finding the corresponding comment in the XML.
